# __*TP 2 - Préparation des données avec Python - Jeathusa*__

## **Objectif :**

Nettoyer le dataset `catnat_dirty.csv` et produire un fichier propre `catnat_clean.csv` prêt à être analysé dans Tableau.

<br>

## - __**Problèmes à résoudre**__
__**Le dataset contient volontairement :**__


- ~150 doublons
- Valeurs manquantes supplémentaires
- Incohérences de casse (Asia, ASIA, asia...)
- Espaces parasites
- Variantes d'orthographe (USA, US, United States...)
- `Start Year` en format texte avec erreurs ("2020 AD", "Year 2020")
- Outliers aberrants (décès négatifs, magnitude à 999)

In [94]:
import pandas as pd
import numpy as np

df = pd.read_csv("./data/raw/catnat_dirty.csv")

In [95]:
df.head()

DisNo.                   Country    Region           Subregion  \
0  2020-0121-IDN                 Indonesia     Asia   South-eastern Asia   
1  2008-0524-USA  United States of America  Americas    Northern America   
2  2015-0281-SLB          Solomon Islands    Oceania           Melanesia   
3  2007-0274-KHM                 Cambodia       Asia  South-eastern Asia   
4  2008-0275-JPN                     Japon      Asia        Eastern Asia   

  Disaster Type    Disaster Subtype Disaster Subgroup              Event Name  \
0         FLOOD     Flood (General)      Hydrological                     NaN   
1      WILDFIRE  Wildfire (General)    Climatological                     NaN   
2         storm    Tropical cyclone    Meteorological  Tropical cylone Raquel   
3      epidemic       Viral disease        Biological                  Dengue   
4    earthquake     Ground movement       Geophysical                     NaN   

  Start Year  Start Month  Total Deaths  No. Injured  Total Affected  \
0       2020          3.0           1.0          NaN         56488.0   
1       2008         11.0           NaN         20.0         55020.0   
2       2015          7.0           9.0          NaN           400.0   
3       2007          7.0         182.0          NaN         17000.0   
4       2008          7.0           1.0        200.0           470.0   

   No. Homeless  Total Damage ('000 US$)  Magnitude  Latitude  Longitude  
0           NaN                      NaN        NaN       NaN        NaN  
1           NaN                2000000.0        NaN       NaN        NaN  
2         400.0                   2000.0        NaN       NaN        NaN  
3           NaN                      NaN        NaN       NaN        NaN  
4           NaN                 110000.0        6.8    39.802    141.464

<hr>

<br>
<br>

## __**Exercice 1 — Exploration et diagnostic**__

### **Rappel**
>
> Avant de modifier, toujours explorer :
>
> - `df.shape` → dimensions
> - `df.info()` → types et nulls
> - `df.describe()` → statistiques
> - `df.isnull().sum()` → comptage des nulls
> - `df.duplicated().sum()` → comptage des doublons
> - `df['col'].value_counts()` → valeurs uniques

<br>
<br>

### __*À faire*__

1. Affichez les dimensions du dataset
2. Affichez les types de données avec `info()`
3. Comptez les valeurs manquantes par colonne (nombre et pourcentage)
4. Comptez le nombre de doublons
5. Affichez les valeurs uniques de `Region` — repérez les incohérences
6. Affichez les statistiques de `Total Deaths` — repérez les anomalies

<br>
<br>

#### __**1. Affichez les dimensions du dataset**__

In [96]:
df.shape
print(f"Number de lignes: {df.shape[0]}")
print(f"Number de la columns: {df.shape[1]}")

Number de lignes: 17510
Number de la columns: 18


In [97]:
df.describe()

Start Month  Total Deaths   No. Injured  Total Affected  No. Homeless  \
count  16599.000000  1.259300e+04  4.507000e+03    1.295100e+04  2.521000e+03   
mean       6.442798  2.751663e+03  2.538213e+03    6.873218e+05  7.226349e+04   
std        3.387116  6.577989e+04  3.230524e+04    7.343419e+06  5.158677e+05   
min        1.000000 -3.600000e+04  1.000000e+00    1.000000e+00  3.000000e+00   
25%        4.000000  5.000000e+00  1.300000e+01    7.000000e+02  5.020000e+02   
50%        7.000000  1.800000e+01  5.000000e+01    6.000000e+03  3.000000e+03   
75%        9.000000  6.100000e+01  2.000000e+02    6.000000e+04  1.700000e+04   
max       12.000000  3.700000e+06  1.800000e+06    3.300000e+08  1.585000e+07   

       Total Damage ('000 US$)     Magnitude     Latitude    Longitude  
count             5.630000e+03  5.249000e+03  2821.000000  2821.000000  
mean              8.448720e+05  3.780406e+04    18.587746    39.675560  
std               5.077550e+06  2.331585e+05    21.591603    77.230791  
min               2.000000e+00 -5.700000e+01   -72.640000  -178.252000  
25%               1.000000e+04  6.900000e+00     3.295000     0.710000  
50%               7.000000e+04  1.400000e+02    23.027000    49.780000  
75%               3.730000e+05  6.851660e+03    36.623000   102.680000  
max               2.100000e+08  1.302587e+07    67.930000   179.650000


#### __**2. Affichez les types de données avec `info()`**__

In [98]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 17510 entries, 0 to 17509
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   DisNo.                   17510 non-null  str    
 1   Country                  17510 non-null  str    
 2   Region                   17161 non-null  str    
 3   Subregion                17510 non-null  str    
 4   Disaster Type            17510 non-null  str    
 5   Disaster Subtype         17510 non-null  str    
 6   Disaster Subgroup        17510 non-null  str    
 7   Event Name               3975 non-null   str    
 8   Start Year               17510 non-null  str    
 9   Start Month              16599 non-null  float64
 10  Total Deaths             12593 non-null  float64
 11  No. Injured              4507 non-null   float64
 12  Total Affected           12951 non-null  float64
 13  No. Homeless             2521 non-null   float64
 14  Total Damage ('000 US$)  5630 non


#### __**3. Comptez les valeurs manquantes par colonne (nombre et pourcentage)**__

In [99]:
manquant = df.isnull().sum()
manquant_pourcentage = (manquant / len(df)) * 100
manquant_df = pd.DataFrame({"Manquant": manquant, "Pourcentage": manquant_pourcentage})
manquant_df["Pourcentage"] = manquant_df["Pourcentage"].round(2).astype(str) + "%"
print(manquant_df)

                         Manquant Pourcentage
DisNo.                          0        0.0%
Country                         0        0.0%
Region                        349       1.99%
Subregion                       0        0.0%
Disaster Type                   0        0.0%
Disaster Subtype                0        0.0%
Disaster Subgroup               0        0.0%
Event Name                  13535       77.3%
Start Year                      0        0.0%
Start Month                   911        5.2%
Total Deaths                 4917      28.08%
No. Injured                 13003      74.26%
Total Affected               4559      26.04%
No. Homeless                14989       85.6%
Total Damage ('000 US$)     11880      67.85%
Magnitude                   12261      70.02%
Latitude                    14689      83.89%
Longitude                   14689      83.89%


#### __**4. Comptez le nombre de doublons**__

In [100]:
n_doublons = df.duplicated().sum()
print(f"Nombre de doublons : {n_doublons}")

Nombre de doublons : 150


#### __**5. Affichez les valeurs uniques de `Region` - repérez les incohérences**__

In [101]:
raw_unique = df['Region'].dropna().unique()
print("Valeurs uniques (brutes):")
print(raw_unique)


Valeurs uniques (brutes):
<StringArray>
[     'Asia ',   'Americas',    'Oceania',       'Asia',       'ASIA',
 ' Americas ',     'Africa',    'Europe ',     'Europe',  ' Americas',
     'EUROPE',  'AMERICAS ',   'americas',   'AMERICAS',     ' Asia ',
    ' Africa',  'Americas ',       'asia',   ' Oceania',      'ASIA ',
    'oceania', ' americas ',      ' Asia',      'asia ',    'OCEANIA',
     'africa',     'AFRICA',     'europe',    'europe ', ' AMERICAS ',
  'americas ',  ' AMERICAS',   ' Europe ',    ' EUROPE',      ' ASIA',
   ' Africa ',  ' americas',    'africa ',      ' asia',    ' africa',
   ' AFRICA ',    'Africa ',    ' AFRICA',   'Oceania ',     ' asia ',
    'AFRICA ',   ' europe ',    'EUROPE ',    ' Europe',   ' africa ',
     ' ASIA ',    ' europe',  ' Oceania ',   'OCEANIA ',  ' OCEANIA ',
   ' EUROPE ',   'oceania ',   ' OCEANIA',  ' oceania ',   ' oceania']
Length: 60, dtype: str


In [102]:
print("\nComptage brut:\n")
print(df['Region'].value_counts(dropna=False))


Comptage brut:

Region
Asia         3754
Americas     2359
Africa       1688
Europe       1201
ASIA          792
             ... 
OCEANIA         8
 OCEANIA        7
 oceania        7
 OCEANIA        5
 oceania        5
Name: count, Length: 61, dtype: int64


<br>

__**`- J'ai fait test sur les variantes sur la colonne Region pour montrer les incohérences (espaces, casse, orthographe).`**__

In [103]:
norm = df['Region'].astype(str).str.strip().str.lower()
print("\nCompte (normalisé: strip + lower):")
print(norm.value_counts())


Compte (normalisé: strip + lower):
Region
asia        6881
americas    4283
africa      3169
europe      2131
oceania      697
Name: count, dtype: int64


#### __**6. Affichez les statistiques de `Total Deaths` - repérez les anomalies**__

In [104]:
totale_mort = df['Total Deaths'].describe()

print("\nStatistiques descriptives pour 'Total Deaths':\n")
print(totale_mort)


Statistiques descriptives pour 'Total Deaths':

count    1.259300e+04
mean     2.751663e+03
std      6.577989e+04
min     -3.600000e+04
25%      5.000000e+00
50%      1.800000e+01
75%      6.100000e+01
max      3.700000e+06
Name: Total Deaths, dtype: float64


 __**`- Comme on peut voir les valeurs min est -3.600000e+04, ce qui est anormal.`**__

In [105]:
# Valeurs négatives dans 'Total Deaths'
neg = df[df['Total Deaths'] < 0]
print("\nNombre de négatifs:", len(neg))
print("\nExemples de valeurs négatives:\n", neg[['Region', 'Total Deaths']].head())


Nombre de négatifs:

 67

Exemples de valeurs négatives:
        Region  Total Deaths
33   Americas          -2.0
136    Africa         -30.0
293      Asia          -1.0
607     Asia          -21.0
754  Americas         -61.0


<br>

## __**Questions**__

**1. Combien y a-t-il de doublons ?**

* D'Après le résultat de `df.duplicated().sum()`, il y a 150 doublons dans le dataset.    
Le dataset contient 17510 lignes, 18 colonnes.

<br>

**2. Quelles colonnes ont le plus de valeurs manquantes ?**

* D'Après le résultat de `manquant = df.isnull().sum()`, les colonnes avec le plus de valeurs manquantes sont :
* No. Homeless                14989     -   (85.6%)
* Latitude                    14689   -    (83.89%)
* Longitude                   14689   -    (83.89%)
* Event Name                  13535    -    (77.3%)
* No. Injured                 13003   -    (74.26%)


<br>

**3. Combien de variantes différentes pour "Asia" ?**
* D'Après le résultat de `df['Region'].value_counts()`, il y a 3 variantes différentes pour "Asia" : "Asia", "ASIA", "asia" qui représentent respectivement 3754 occurences.

<br>

**4. Y a-t-il des valeurs négatives dans `Total Deaths` ?**
* D'Après le résultat de `df['Total Deaths'].describe()`, il y a des valeurs négatives dans `Total Deaths` avec une valeur minimum de -36000, ce qui est anormal pour ce type de données.

<br>

<br>
<hr>
<br>

## __**Exercice 2 — Suppression des doublons**__

> **Rappel**
>
> ```python
> # Identifier les doublons
> df[df.duplicated()]
> df[df.duplicated(keep=False)]  # Inclut les originaux
>
> # Supprimer les doublons
> df = df.drop_duplicates()
> df = df.drop_duplicates(subset=['col1', 'col2'])  # Sur certaines colonnes
> ```

<br>
<br>

### __*À faire*__

1. Affichez quelques lignes dupliquées pour vérifier
2. Supprimez les doublons exacts
3. Vérifiez que les doublons ont bien été supprimés

<br>
<br>


#### __**1. Affichez quelques lignes dupliquées pour vérifier**__

In [106]:
df[df.duplicated(keep=False)]

DisNo.                    Country    Region  \
27     1969-0071-IND                      India      ASIA   
125    2005-0583-USA                        USA  americas   
322    1963-0055-BEL                    Belgium    Europe   
371    2023-0510-MNG                   Mongolia      Asia   
456    1996-0226-CHN                      China      Asia   
...              ...                        ...       ...   
17194  2023-0290-SLE               Sierra Leone    Africa   
17228  1997-0126-SLV                El Salvador  Americas   
17339  2010-0053-PAK                   Pakistan      ASIA   
17410  2013-0286-USA   United States of America  Americas   
17413  2015-0226-IND                      India     asia    

                             Subregion        Disaster Type  Disaster Subtype  \
27                       Southern Asia                Storm  Tropical cyclone   
125                   Northern America                Flood    Riverine flood   
322                     Western Europe  Extreme Temperature         Cold wave   
371                       Eastern Asia                Flood       Flash flood   
456                       Eastern Asia                Storm  Tropical cyclone   
...                                ...                  ...               ...   
17194               Sub-Saharan Africa                STORM   Storm (General)   
17228  Latin America and the Caribbean                STORM  Tropical cyclone   
17339                    Southern Asia                Flood    Riverine flood   
17410                 Northern America                Flood    Riverine flood   
17413                    Southern Asia  Mass movement (wet)   Landslide (wet)   

      Disaster Subgroup Event Name Start Year  Start Month  Total Deaths  \
27       Meteorological        NaN       1969          5.0      600000.0   
125        Hydrological        NaN       2005         10.0          11.0   
322      Meteorological        NaN       1963          NaN          12.0   
371        Hydrological        NaN       2023          8.0           4.0   
456      Meteorological     Willie       1996          9.0          38.0   
...                 ...        ...        ...          ...           ...   
17194    Meteorological        NaN       2023          5.0          15.0   
17228    Meteorological     Andres       1997          6.0           4.0   
17339      Hydrological        NaN       2010          2.0          22.0   
17410      Hydrological        NaN       2013          5.0           3.0   
17413      Hydrological        NaN       2015          6.0           3.0   

       No. Injured  Total Affected  No. Homeless  Total Damage ('000 US$)  \
27             NaN        260000.0           NaN                   8330.0   
125            NaN          3000.0           NaN                      NaN   
322            NaN             NaN           NaN                      NaN   
371            NaN          1230.0           NaN                      NaN   
456            NaN             NaN           NaN                 100000.0   
...            ...             ...           ...                      ...   
17194         17.0            17.0           NaN                      NaN   
17228          NaN          2000.0        2000.0                      NaN   
17339          NaN             NaN           NaN                      NaN   
17410          NaN           300.0           NaN                   2000.0   
17413          NaN          9000.0           NaN                      NaN   

       Magnitude  Latitude  Longitude  
27           NaN       NaN        NaN  
125     38290.00       NaN        NaN  
322       -22.00       NaN        NaN  
371          NaN       NaN        NaN  
456       130.00       NaN        NaN  
...          ...       ...        ...  
17194        NaN       NaN        NaN  
17228        NaN       NaN        NaN  
17339        NaN       NaN        NaN  
17410   33907.85   28.8990     -98.89  
17413        NaN   26.9195    

<br>
<br>

#### __**2. Supprimez les doublons exacts**__

In [107]:
df = df.drop_duplicates()

<br>
<br>

#### __**3. Vérifiez que les doublons ont bien été supprimés**__

In [108]:
n_doublons = df.duplicated().sum()
print(f"Nombre de doublons : {n_doublons}")

Nombre de doublons : 0


<br>
<hr>
<br>

## __**Exercice 3 — Correction des types**__

> **Rappel**
>
> ```python
> # Vérifier les types
> df.dtypes
>
> # Convertir en numérique (erreurs → NaN)
> df['col'] = pd.to_numeric(df['col'], errors='coerce')
>
> # Convertir en entier
> df['col'] = df['col'].astype(int)
>
> # Convertir en date
> df['col'] = pd.to_datetime(df['col'], errors='coerce')
> ```

```python
# petit hint...
dtype.name existe !

# AD signifie Anno Domini, elle correspond à la formule français après JC ...
```

<br>
<br>

### __*À faire*__

1. Vérifiez le type de `Start Year`
2. Affichez quelques valeurs problématiques (contenant "Year" ou "AD")
3. Nettoyez la colonne : supprimez le texte et convertissez en numérique
4. Vérifiez le résultat

<br>
<br>

#### __**1. Vérifiez le type de `Start Year`**__

In [109]:
print(f"Type de la colonne 'Start Year': {df['Start Year'].dtype}")

Type de la colonne 'Start Year': str


<br>
<br>

#### __**2. Affichez quelques valeurs problématiques (contenant "Year" ou "AD")**__

In [110]:
valeurs_problematiques = df[df['Start Year'].astype(str).str.contains("Year|AD", na=False)]
print("Exemples de valeurs problématiques :")
print(valeurs_problematiques[['Start Year']].head())

Exemples de valeurs problématiques :
    Start Year
26     1982 AD
138  Year 2012
480    2021 AD
506  Year 2017
634  Year 2003


<br>
<br>

#### __**3. Nettoyez la colonne : supprimez le texte et convertissez en numérique**__

In [111]:
# Supprimer les text chaînes de caractères "Year" et "AD"
df['Start Year'] = df['Start Year'].astype(str).str.replace(r'Year|AD', '', regex=True).str.strip()

# Convertir la colonne en numérique
df['Start Year'] = pd.to_numeric(df['Start Year'], errors='coerce')

<br>
<br>

#### __**4. Vérifiez le résultat**__

In [112]:
print(f"Nouveau type de la colonne 'Start Year': {df['Start Year'].dtype}")

# Vérification des valeurs textuelles restantes
valeurs_texte_restantes = df['Start Year'].apply(lambda x: isinstance(x, str)).sum()
print(f"Nombre de valeurs textuelles restantes : {valeurs_texte_restantes}")

print("\nValeurs de la colonne nettoyée :")
print(df['Start Year'].head())

Nouveau type de la colonne 'Start Year': int64
Nombre de valeurs textuelles restantes : 0

Valeurs de la colonne nettoyée :
0    2020
1    2008
2    2015
3    2007
4    2008
Name: Start Year, dtype: int64


<br>
<hr>
<br>

## __**Exercice 4 — Traitement des valeurs manquantes**__

> **Rappel**
>
> | Situation                               | Action                             |
> | --------------------------------------- | ---------------------------------- |
> | Peu de nulls, colonne critique          | Supprimer les lignes               |
> | Beaucoup de nulls, colonne non critique | Supprimer la colonne ou garder NaN |
> | Numérique, distribution asymétrique     | Imputer par médiane                |
> | Catégoriel                              | Imputer par mode ou "Inconnu"      |
>
> ```python
> df.dropna(subset=['col'])           # Supprimer lignes
> df['col'].fillna(valeur)            # Remplacer
> df['col'].fillna(df['col'].median())  # Médiane
> ```

<br>
<br>

### __**À faire**__

1. Pour `Region` (peu de nulls) : supprimez les lignes avec null
2. Pour `Event Name` : remplacez les nulls par "Non nommé"
3. Pour `Total Deaths` : décidez d'une stratégie et appliquez-la
4. Pour `Magnitude` : laissez les nulls (n'a pas de sens pour tous les types)
5. Vérifiez le résultat

<br>
<br>

#### __**1. Pour `Region` (peu de nulls) : supprimez les lignes avec null**__


In [113]:

print(f"Nombre de lignes dans 'Region': {len(df)}")
print(f"Nuls dans 'Region' : {df['Region'].isnull().sum()}")

Nombre de lignes dans 'Region': 17360
Nuls dans 'Region' : 346


In [114]:
# Suppression des lignes où Region est nulle
df.dropna(subset=['Region'], inplace=True)

In [115]:
# Après la suppression
print(f"Nombre de lignes après suppression: {len(df)}")
print(f"Nuls dans 'Region' après: {df['Region'].isnull().sum()}")

Nombre de lignes après suppression: 17014
Nuls dans 'Region' après: 0


<br>
<br>

#### __**2. Pour `Event Name` : remplacez les nulls par "Non nommé"**__


In [116]:
df['Event Name'].fillna('Non nommé', inplace=True)

C:\Users\jeath\AppData\Local\Temp\ipykernel_17288\3554683477.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df['Event Name'].fillna('Non nommé', inplace=True)


0                       Non nommé
1                       Non nommé
2          Tropical cylone Raquel
3                          Dengue
4                       Non nommé
                   ...           
17505                   Non nommé
17506     Tropical storm "Emily" 
17507                   Non nommé
17508                   Non nommé
17509                   Non nommé
Name: Event Name, Length: 17014, dtype: str

<br>
<br>

#### __**3. Pour `Total Deaths` : décidez d'une stratégie et appliquez-la**__


In [117]:
# Calculer la médiane des décès
median_deaths = df['Total Deaths'].median()
print(f"La médiane des décès est : {median_deaths}")

# Remplacer les nuls par la médiane
df['Total Deaths'].fillna(median_deaths, inplace=True)

La médiane des décès est : 18.0


C:\Users\jeath\AppData\Local\Temp\ipykernel_17288\736382464.py:6: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df['Total Deaths'].fillna(median_deaths, inplace=True)


0          1.0
1         18.0
2          9.0
3        182.0
4          1.0
         ...  
17505     18.0
17506      3.0
17507      7.0
17508     50.0
17509     12.0
Name: Total Deaths, Length: 17014, dtype: float64

<br>
<br>

#### __**4. Pour `Magnitude` : laissez les nulls (n'a pas de sens pour tous les types)**__


In [118]:
print(df['Magnitude'].dtype)

float64


In [119]:
n_null = df['Magnitude'].isnull().sum()
print(f"Nulls: {n_null}/{len(df)} ({n_null/len(df)*100:.2f}%)")
print(df['Magnitude'].value_counts(dropna=False).head(10))

Nulls: 11929/17014 (70.11%)
Magnitude
NaN      11929
100.0       85
120.0       67
6.0         66
150.0       66
6.3         66
6.4         65
6.5         63
5.6         63
130.0       63
Name: count, dtype: int64


In [120]:
mag_num = pd.to_numeric(df['Magnitude'], errors='coerce')
print(mag_num.describe())

count    5.085000e+03
mean     3.788548e+04
std      2.355306e+05
min     -5.700000e+01
25%      6.900000e+00
50%      1.400000e+02
75%      6.800000e+03
max      1.302587e+07
Name: Magnitude, dtype: float64


__**`- Je l’ai laissé en `NaN` car, à mon avis personnel, la magnitude n’est pas pertinente pour tous les types d’événements, imputer une valeur introduirait une fausse précision. Cette décision est documentée et on peut ajouter une colonne indicatrice (`exemple:  has_magnitude;`) ou traiter séparément les événements avec magnitude.`**__

<br>
<br>

#### __**5. Vérifiez le résultat**__


In [121]:
print("Nombre de valeurs manquantes par colonne après traitement :")
print(df.isnull().sum())

Nombre de valeurs manquantes par colonne après traitement :
DisNo.                         0
Country                        0
Region                         0
Subregion                      0
Disaster Type                  0
Disaster Subtype               0
Disaster Subgroup              0
Event Name                 13139
Start Year                     0
Start Month                  890
Total Deaths                4774
No. Injured                12643
Total Affected              4441
No. Homeless               14570
Total Damage ('000 US$)    11540
Magnitude                  11929
Latitude                   14289
Longitude                  14289
dtype: int64


<br>
<hr>
<br>

## __**Exercice 5 — Traitement des outliers**__

> **Rappel**
>
> Deux questions :
>
> 1. Est-ce une erreur ? → Corriger ou supprimer
> 2. Est-ce une valeur extrême réelle ? → Garder (ou analyser séparément)
>
> ```python
> # Valeurs impossibles
> df[df['col'] < 0]
> df[df['col'] > seuil_max]
>
> # Supprimer
> df = df[df['col'] >= 0]
>
> # Capper
> df['col'] = df['col'].clip(lower=0, upper=max_val)
> ```

<br>
<br>

### __*À faire*__

1. Identifiez les valeurs négatives dans `Total Deaths`
2. Identifiez les valeurs aberrantes dans `Magnitude` (> 10)
3. Décidez : supprimer ou corriger ?
4. Appliquez le traitement
5. Vérifiez le résultat

<br>
<br>

#### __**1. Identifiez les valeurs négatives dans `Total Deaths`**__

In [122]:
print((df['Total Deaths'] < 0).sum())
print(df[df['Total Deaths'] < 0][['Event Name','Region','Total Deaths']].head())

65
    Event Name    Region  Total Deaths
33     Isidore  Americas          -2.0
136        NaN    Africa         -30.0
293        NaN      Asia          -1.0
607        NaN     Asia          -21.0
754        NaN  Americas         -61.0


<br>
<br>

#### __**2. Identifiez les valeurs aberrantes dans `Magnitude` (> 10)`**__

In [123]:
mag = pd.to_numeric(df['Magnitude'], errors='coerce')
print((mag > 10).sum())
print(df[mag > 10][['Disaster Type','Event Name','Magnitude']].head())

3364
   Disaster Type Event Name  Magnitude
5          flood        NaN   365800.0
32         Flood        NaN      200.0
34      Wildfire        NaN     1320.0
40         Flood        NaN   444500.0
42         STORM        NaN      180.0


<br>
<br>

#### __**3. Décidez : supprimer ou corriger ?**__

**Décès négatifs :**
Ces valeurs sont mathématiquement et logiquement impossibles (on ne peut pas avoir un nombre de morts négatif). Il s'agit clairement d'erreurs de saisie. `Plutôt que de supprimer ces lignes et perdre d'autres informations utiles (région, type de catastrophe, magnitude), j'ai choisi de les remplacer par NaN`, puis de les imputer avec la médiane. La médiane est préférable à la moyenne car elle est robuste aux valeurs extrêmes et ne distorsionne pas la distribution des données en cas de valeurs aberrantes.

<br>

**Magnitudes > 10 ou = 999 :**
Une magnitude supérieure à 10 est physiquement impossible. La valeur 999 est clairement un code d'erreur. Ces données sont inutilisables pour l'analyse. Cependant, `j'ai remplacé ces valeurs par NaN plutôt que de supprimer les lignes entières, car les autres variables (région, date, nombre de morts) restent valides et exploitables`. Cela maximise la rétention d'informations tout en garantissant la qualité de l'analyse.

<br>

`En résumé : Ces choix visent à concilier qualité des données et conservation d'informations utiles.`

<br>
<br>

#### __**4. Appliquez le traitement**__

In [124]:
# Identification et remplacement les décès négatifs par NaN
print("Décès négatifs détectés :", (df['Total Deaths'] < 0).sum())
df.loc[df['Total Deaths'] < 0, 'Total Deaths'] = np.nan

# Imputer avec la médiane
median_deaths = df['Total Deaths'].median()
df['Total Deaths'].fillna(median_deaths, inplace=True)
print(f"Imputés avec la médiane : {median_deaths}")
print(f"Décès négatifs restants : {(df['Total Deaths'] < 0).sum()}\n")


Décès négatifs détectés : 65
Imputés avec la médiane : 18.0
Décès négatifs restants : 0



C:\Users\jeath\AppData\Local\Temp\ipykernel_17288\2847685292.py:7: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df['Total Deaths'].fillna(median_deaths, inplace=True)


In [125]:

mag = pd.to_numeric(df['Magnitude'], errors='coerce')

# Identifier les magnitudes > 10 ou == 999
print("Magnitudes > 10 détectées :", (mag > 10).sum())
print("Magnitudes == 999 détectées :", (df['Magnitude'] == 999).sum())

# Remplacement par NaN
df.loc[(mag > 10) | (df['Magnitude'] == 999), 'Magnitude'] = np.nan

print(f"Magnitudes invalides remplacées par NaN")
print(f"Magnitudes > 10 restantes : {(pd.to_numeric(df['Magnitude'], errors='coerce') > 10).sum()}\n")


Magnitudes > 10 détectées : 3364
Magnitudes == 999 détectées : 22
Magnitudes invalides remplacées par NaN
Magnitudes > 10 restantes : 0



<br>
<br>

#### __**5. Vérifiez le résultat**__

In [126]:
print("État final du dataset :")
print(f"Décès manquants (NaN) : {df['Total Deaths'].isna().sum()}")
print(f"Magnitudes manquantes (NaN) : {df['Magnitude'].isna().sum()}")
print(f"Forme du dataset : {df.shape}")

État final du dataset :
Décès manquants (NaN) : 4839
Magnitudes manquantes (NaN) : 15293
Forme du dataset : (17014, 18)


<br>
<hr>
<br>

## __**Exercice 6 — Nettoyage du texte**__

> **Rappel**
>
> ```python
> # Casse
> df['col'].str.lower()
> df['col'].str.upper()
> df['col'].str.title()
>
> # Espaces
> df['col'].str.strip()
>
> # Remplacement
> df['col'].replace({'old': 'new'})
>
> # Pipeline complet
> df['col'] = df['col'].str.strip().str.lower()
> ```

### __*À faire*__

1. Nettoyez `Region` : strip + title case
2. Vérifiez avec `value_counts()` — combien de catégories maintenant ?
3. Nettoyez `Disaster Type` de la même manière
4. Nettoyez `Country` : strip + title case
5. Corrigez les variantes de pays (USA → United States, etc.)

<br>
<br>

#### __**1. Nettoyez `Region` : strip + title case**__

In [127]:
df['Region'] = df['Region'].astype(str).str.strip().str.title()

<br>
<br>

#### __**2. Vérifiez avec `value_counts()` — combien de catégories maintenant ?**__

In [128]:
print(df['Region'].value_counts())
print("\n\nRegion après nettoyage :")
print(f"Nombre de catégories uniques : {df['Region'].nunique()}\n")

Region
Asia        6817
Americas    4243
Africa      3142
Europe      2117
Oceania      695
Name: count, dtype: int64


Region après nettoyage :
Nombre de catégories uniques : 5



<br>
<br>

#### __**3. Nettoyez `Disaster Type` de la même manière**__

In [129]:
df['Disaster Type'] = df['Disaster Type'].astype(str).str.strip().str.title()
print(df['Disaster Type'].value_counts().head())

Disaster Type
Flood                  5945
Storm                  4790
Earthquake             1584
Epidemic               1488
Mass Movement (Wet)     836
Name: count, dtype: int64


<br>
<br>

#### __**4. Nettoyez `Country` : strip + title case**__

In [130]:
df['Country'] = df['Country'].fillna('').astype(str).str.strip().replace('', np.nan).str.title()
print(df['Country'].value_counts().head())

Country
United States Of America    906
China                       854
India                       711
Indonesia                   560
Philippines                 559
Name: count, dtype: int64


<br>
<br>

#### __**5. Corrigez les variantes de pays (USA → United States, etc.)**__

In [131]:
df['Country'] = df['Country'].str.replace('Usa', 'United States')
df['Country'] = df['Country'].str.replace('Us', 'United States') 
df['Country'] = df['Country'].str.replace('Uk', 'United Kingdom')

In [132]:
print("\nAprès correction :")
print(df['Country'].value_counts().head(10))
print(f"Nombre de pays : {df['Country'].nunique()}")


Après correction :
Country
United States Of America    906
China                       854
India                       711
Indonesia                   560
Philippines                 559
Japan                       347
Mexico                      298
Bangladesh                  283
Pakistan                    257
Viet Nam                    256
Name: count, dtype: int64
Nombre de pays : 287


<br>
<hr>
<br>


## __**Exercice 7 — Création de colonnes utiles**__

> **Rappel**
>
> ```python
> # Nouvelle colonne calculée
> df['new'] = df['col1'] / df['col2']
>
> # Colonne à partir d'une autre
> df['Decennie'] = (df['Year'] // 10) * 10
>
> # Extraction de date
> df['Annee'] = df['Date'].dt.year
> df['Mois'] = df['Date'].dt.month
> ```

### __*À faire*__

1. Créez une colonne `Decennie` à partir de `Start Year`
2. Créez une colonne `Has_Deaths` (booléen : True si Total Deaths > 0)
3. Vérifiez vos nouvelles colonnes

<br>
<br>

#### __**1. Créez une colonne `Decennie` à partir de `Start Year`**__

In [133]:
df['Decennie'] = (df['Start Year'] // 10) * 10
print(df[['Start Year', 'Decennie']].head())

   Start Year  Decennie
0        2020      2020
1        2008      2000
2        2015      2010
3        2007      2000
4        2008      2000


-   ` df['Start Year'] // 10` : Division entière par 10 (enlève le dernier chiffre)
-   Exemple : `2005 ---> 200`, `1995 ---> 199`
-   `*10` : Multiplie par 10 pour obtenir la décennie
-   Exemple : 200 * 10 = 2000, 199 * 10 = 1990

**Résultat** : Regroupe les années par décennie (`2000-2009, 1990-1999, etc.....`)

<br>
<br>

#### __**2. Créez une colonne `Has_Deaths` (booléen : True si Total Deaths > 0)**__

In [134]:
df['Has_Deaths'] = df['Total Deaths'] > 0
print(df[['Total Deaths', 'Has_Deaths']].head())

   Total Deaths  Has_Deaths
0           1.0        True
1           NaN       False
2           9.0        True
3         182.0        True
4           1.0        True



-   **Résultat** :
-   `True` si Total Deaths > 0 (`il y a eu des décès`)
-   `False` si Total Deaths = 0 (`pas de décès`)

<br>
<br>

#### __**3. Vérifiez vos nouvelles colonnes**__

In [135]:
print(df[['Start Year', 'Decennie', 'Total Deaths', 'Has_Deaths']].head(10))
print("\nData types:")
print(df[['Decennie', 'Has_Deaths']].dtypes)

   Start Year  Decennie  Total Deaths  Has_Deaths
0        2020      2020           1.0        True
1        2008      2000           NaN       False
2        2015      2010           9.0        True
3        2007      2000         182.0        True
4        2008      2000           1.0        True
5        1997      1990          16.0        True
6        1999      1990         325.0        True
7        2004      2000           NaN       False
8        1961      1960         275.0        True
9        1914      1910          34.0        True

Data types:
Decennie      int64
Has_Deaths     bool
dtype: object


<br>
<hr>
<br>

## __**Exercice 8 — Sélection et export**__

> **Rappel**
>
> ```python
> # Sélectionner des colonnes
> df_export = df[['col1', 'col2', 'col3']]
>
> # Renommer
> df_export = df_export.rename(columns={'old': 'new'})
>
> # Export
> df_export.to_csv("fichier.csv", index=False)
> ```

## __*À faire*__

1. Sélectionnez les colonnes utiles pour l'analyse
2. Renommez les colonnes si nécessaire (noms clairs, sans espaces)
3. Faites une vérification finale avec `info()` et `head()`
4. Exportez en CSV : `catnat_clean.csv`

<br>
<br>

#### __**1. Sélectionnez les colonnes utiles pour l'analyse**__

In [136]:
df_export = df[['DisNo.', 'Disaster Type', 'Region', 'Country', 
               'Total Deaths', 'Start Year', 'Event Name', 'Magnitude', 
               'Decennie', 'Has_Deaths']]
df_export.head()

DisNo. Disaster Type    Region                   Country  \
0  2020-0121-IDN         Flood      Asia                 Indonesia   
1  2008-0524-USA      Wildfire  Americas  United States Of America   
2  2015-0281-SLB         Storm   Oceania           Solomon Islands   
3  2007-0274-KHM      Epidemic      Asia                  Cambodia   
4  2008-0275-JPN    Earthquake      Asia                     Japon   

   Total Deaths  Start Year              Event Name  Magnitude  Decennie  \
0           1.0        2020                     NaN        NaN      2020   
1           NaN        2008                     NaN        NaN      2000   
2           9.0        2015  Tropical cylone Raquel        NaN      2010   
3         182.0        2007                  Dengue        NaN      2000   
4           1.0        2008                     NaN        6.8      2000   

   Has_Deaths  
0        True  
1       False  
2        True  
3        True  
4        True

<br>
<br>

#### __**2. Renommez les colonnes si nécessaire (noms clairs, sans espaces)**__

In [137]:
df_export = df_export.rename(columns={
    'DisNo.': 'ID_Catastrophe',
    'Disaster Type': 'Type_Catastrophe',
    'Start Year': 'Annee_Debut',
    'Event Name': 'Nom_Evenement',
    'Total Deaths': 'Nombre_Deces',
    'Has_Deaths': 'A_Des_Deces',
    'Country': 'Pays'
})

<br>
<br>

#### __**3. Faites une vérification finale avec `info()` et `head()`**__

In [138]:
print("Informations sur le dataset :")
df_export.info()

print("\n\n\nPremiers enregistrements :")
df_export.head()

Informations sur le dataset :
<class 'pandas.DataFrame'>
Index: 17014 entries, 0 to 17509
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ID_Catastrophe    17014 non-null  str    
 1   Type_Catastrophe  17014 non-null  str    
 2   Region            17014 non-null  str    
 3   Pays              17014 non-null  str    
 4   Nombre_Deces      12175 non-null  float64
 5   Annee_Debut       17014 non-null  int64  
 6   Nom_Evenement     3875 non-null   str    
 7   Magnitude         1721 non-null   float64
 8   Decennie          17014 non-null  int64  
 9   A_Des_Deces       17014 non-null  bool   
dtypes: bool(1), float64(2), int64(2), str(5)
memory usage: 1.3 MB



Premiers enregistrements :


ID_Catastrophe Type_Catastrophe    Region                      Pays  \
0  2020-0121-IDN            Flood      Asia                 Indonesia   
1  2008-0524-USA         Wildfire  Americas  United States Of America   
2  2015-0281-SLB            Storm   Oceania           Solomon Islands   
3  2007-0274-KHM         Epidemic      Asia                  Cambodia   
4  2008-0275-JPN       Earthquake      Asia                     Japon   

   Nombre_Deces  Annee_Debut           Nom_Evenement  Magnitude  Decennie  \
0           1.0         2020                     NaN        NaN      2020   
1           NaN         2008                     NaN        NaN      2000   
2           9.0         2015  Tropical cylone Raquel        NaN      2010   
3         182.0         2007                  Dengue        NaN      2000   
4           1.0         2008                     NaN        6.8      2000   

   A_Des_Deces  
0         True  
1        False  
2         True  
3         True  
4         True

<br>
<br>

#### __**4. Exportez en CSV : `catnat_clean.csv`**__

In [139]:
output_path = "data/processed/catnat_clean_2.0.csv"

df_export.to_csv(output_path, index=False)
print(f"\nFichier '{output_path}' exporté avec succès !")


Fichier 'data/processed/catnat_clean_2.0.csv' exporté avec succès !


<br>
<hr>
<br>


## __**Exercice 9 — Vérification dans Tableau**__

<br>

### **Tableau Public :**
https://public.tableau.com/app/profile/jeathusan.kugathas/viz/NettoyageetAnalyse-DonnesdeCatastrophes/Localisationgographiquedescatastrophes


<br>
<br>

## __*À faire*__

1. Ouvrez Tableau Public
2. Connectez-vous au fichier `catnat_clean.csv`
3. Vérifiez dans l'écran "Source de données" :
   - Les types sont-ils corrects ? (# pour nombres, Abc pour texte)
   - Les nombres sont-ils bien reconnus ?
4. Créez un bar chart : `Type_Catastrophe` vs `COUNT(ID_Catastrophe)`
   - Les catégories sont-elles propres ? (pas de doublons)
5. Créez un bar chart : `Region` vs `SUM(Deces)`
   - Les 5 régions sont-elles bien distinctes ?
6. Créez une carte avec `Country`
   - Les pays sont-ils reconnus ?
